src/components/MultiHopTrade/components/SharedTradeInput/SharedRecipientAddress.tsx

```diff
@@ -0,0 +1,387 @@
+            const parseAddressInputWithChainIdArgs = {
+              assetId: buyAssetAssetId,
+              chainId: buyAssetChainId,
+              urlOrAddress: value,
+              disableUrlParsing: true,
+            }
+            const { address } = await parseAddressInputWithChainId(parseAddressInputWithChainIdArgs)
+            const invalidMessage = 'common.invalidAddress'
+            return address ? true : invalidMessage
+          } catch (e) {
+            // This function should never throw, but in case it ever does, we never want to have a stale manual receive address stored
+            console.error(e)
+            onError()
+          }
+        },
+      },
+      minLength: 1,
+    }),
+    [buyAssetAssetId, buyAssetChainId, onError],
+  )
+
+  const handleEditRecipientAddressClick = useCallback(() => {
+    onEdit()
+    setIsRecipientAddressEditing(true)
+  }, [onEdit])
+
+  const handleCancelClick = useCallback(() => {
+    onCancel()
+    setIsRecipientAddressEditing(false)
+    setFormValue(SendFormFields.Input, '')
+  }, [onCancel, setFormValue])
+
+  const resetManualReceiveAddress = useCallback(() => {
+    onReset()
+    // Reset the form value itself, to avoid the user going from
+    // custom recipient -> cleared custom recipient -> custom recipient where the previously set custom recipient
+    // would be displayed, wrongly hinting this is the default wallet address
+    setFormValue(SendFormFields.Input, '')
+  }, [onReset, setFormValue])
+
+  const handleSubmit = useCallback(
+    (values: FieldValues) => {
+      // We don't need to revalidate here as submit will only be enabled if the form is valid
+      const address = values[SendFormFields.Input].trim()
+      onSubmit(address)
+      setIsRecipientAddressEditing(false)
+    },
+    [onSubmit],
+  )
+
+  const handleFormSubmit = useMemo(
+    () => handleFormContextSubmit(handleSubmit),
+    [handleFormContextSubmit, handleSubmit],
+  )
+
+  const forceDisplayManualAddressEntry = useMemo(() => {
+    if (isWalletReceiveAddressLoading) return false
+    if (!isConnected) return false
+    if (isAccountsMetadataLoading && !sellAssetAccountId) return false
+    if (manualReceiveAddress) return false
+    if (!walletReceiveAddress) return true
+    if (!walletSupportsBuyAssetChain) return true
+    // Ledger "supports" all chains, but may not have them connected
+    if (wallet && isLedger(wallet)) return !buyAssetAccountIds.length
+    // We want to display the manual address entry if the wallet doesn't support the buy asset chain
+    if (shouldForceManualAddressEntry) return true
+
+    return false
+  }, [
+    isConnected,
+    isAccountsMetadataLoading,
+    sellAssetAccountId,
+    manualReceiveAddress,
+    walletReceiveAddress,
+    walletSupportsBuyAssetChain,
+    wallet,
+    buyAssetAccountIds.length,
+    shouldForceManualAddressEntry,
+    isWalletReceiveAddressLoading,
+  ])
```

Comment:
Preferably blocking: we have identical logic in `useManuaReceiveAddressIsRequired`, a helper might be useful.

Reply:
Ah yep this was meant to be included, will do!